### PR TITLE
Adding data-param-* to amp-dailymotion

### PR DESCRIPTION
**Pull request by PurpleBabar**

### CODE REVIEW

**Comment by PurpleBabar** [comment]
I signed it!

**Comment by PurpleBabar** [comment]
Hi :) 
Travis ci tells me:
 [✖] https://developer.dailymotion.com/player#player-parameters
ERROR Possible dead link(s) found in extensions/amp-dailymotion/amp-dailymotion.md (please update, or whitelist in build-system/tasks/check-links.js).

The thing is, this link exists and is valid :/ ([see here](https://developer.dailymotion.com/player#player-parameters))

What should i do to be sure, the link passes the tests ?

Thanks in advance:)

**Comment by zeno** [comment]
Thanks @PurpleBabar for this PR!

**Comment by aghassemi** [comment]
Thanks for the PR @PurpleBabar.
/cc @rsimha-amp regarding deadlink report. Raghu, `https://developer.dailymotion.com/player#player-parameters` is being reported as deadlink but it is not. Maybe because links has a `#hash` ?

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js [hunk 1]
```diff
@@ -153,8 +156,9 @@ class AmpDailymotion extends AMP.BaseElement {
     iframe.setAttribute('frameborder', '0');
     iframe.setAttribute('allowfullscreen', 'true');
     dev().assert(this.videoid_);
-    iframe.src = 'https://www.dailymotion.com/embed/video/' +
-     encodeURIComponent(this.videoid_ || '') + '?' + this.getQuery_();
+    iframe.src = this.getIframeSrc_();
+    // iframe.src = 'https://www.dailymotion.com/embed/video/' +
```
Comment (aghassemi) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js]: remove commented out code.

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js [hunk 2]
```diff
@@ -264,11 +274,17 @@ class AmpDailymotion extends AMP.BaseElement {
       'info',
     ];
 
-    settings.forEach(setting => {
-      this.addQueryParam_(setting, query);
+    let explicitParams = dict();
+
+    explicitParamsAttributes.forEach(explicitParam => {
+      this.addDictParam_(explicitParam, explicitParams);
```
Comment (aghassemi) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js]: no need for `addDicParam_` we can do:
1- add `data-` to all the `explicitParamsAttributes` values
2- import `addParamToUrl`
3- inline the whole thing:
```
explicitParamsAttributes.forEach(explicitParam => {
   const val = this.element.getAttribute(explicitParam);
   if (val) {
      addParamToUrl(iframeSrc, explicitParam, val);
   }
});
```

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js [hunk 3]
```diff
@@ -37,6 +37,7 @@ describe('amp-dailymotion', () => {
       }
       if (optCustomSettings) {
         dailymotion.setAttribute('data-start', 123);
+        dailymotion.setAttribute('data-param-origin', 'example.com');
```
Comment (aghassemi) [inline extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js]: maybe use a value that requires ulr encoding?

File: extensions/amp-dailymotion/amp-dailymotion.md [hunk 4]
```diff
@@ -110,6 +110,16 @@ Indicates whether to show video information (title and owner) on the start scree
 * Value: `"true"` or `"false"`
 * Default value: `"true"`
 
+**data-param-*** (optional)
+
+All `data-param-*` attributes will be added as query parameter to the Dailymotion iframe src. This may be used to pass custom values not explicitly declared.
+
+Keys and values  will be URI encoded.
```
Comment (aghassemi) [inline extensions/amp-dailymotion/amp-dailymotion.md]: nit: extra space between `values` and `will`

File: extensions/amp-dailymotion/amp-dailymotion.md [hunk 5]
```diff
@@ -110,6 +110,16 @@ Indicates whether to show video information (title and owner) on the start scree
 * Value: `"true"` or `"false"`
 * Default value: `"true"`
 
+**data-param-*** (optional)
+
+All `data-param-*` attributes will be added as query parameter to the Dailymotion iframe src. This may be used to pass custom values not explicitly declared.
```
Comment (aghassemi) [inline extensions/amp-dailymotion/amp-dailymotion.md]: /cc @bpaduch for documentation review

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js (hunk 1 above)
Comment (PurpleBabar) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js]: Sorry for the oversight

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js (hunk 2 above)
Comment (PurpleBabar) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js]: I was trying not to change too much the previous code :p works for me :)

File: extensions/amp-dailymotion/amp-dailymotion.md (hunk 4 above)
Comment (PurpleBabar) [inline extensions/amp-dailymotion/amp-dailymotion.md]: sorry for the oversight

File: extensions/amp-dailymotion/amp-dailymotion.md (hunk 5 above)
Comment (bpaduch) [inline extensions/amp-dailymotion/amp-dailymotion.md]: Rewrite as: 

All `data-param-*` attributes are added as query parameters to the `src` value of the embedded Dailymotion iframe. You can use this attribute to pass custom values not explicitly declared.

**Comment by rsimha-amp** [comment]
@aghassemi, @PurpleBabar, if you're sure the link works as typed, you could try removing the page anchor to see if that's the issue. You could also run the tests locally via `gulp check-links --files <your md file>` to debug. Either way, some links are flaky when accessed from Travis, in which case you can add them to the whitelist in `tasks/check-links.js`.

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js (hunk 3 above)
Comment (PurpleBabar) [inline extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js]: Isn't a domain name only ASCII and dots (which is safe for escaping) ?

**Comment by PurpleBabar** [comment]
I wanted to avoid passing by the whitelist for the link :/ will definitely try without the anchor though ;)

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js [hunk 6]
```diff
@@ -24,6 +24,10 @@ import {
 import {getData, listen} from '../../../src/event-helper';
 import {videoManagerForDoc} from '../../../src/services';
 import {parseQueryString} from '../../../src/url';
+import {getDataParamsFromAttributes} from '../../../src/dom';
+import {addParamsToUrl} from '../../../src/url';
```
Comment (aghassemi) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js]: please merge the two imports (`import {foo, bar} from 'baz'`)

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js (hunk 3 above)
Comment (aghassemi) [inline extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js]: yeah but for the purpose of testing, it doesn't matter, would be nice to have a test that ensure proper encoding.

Comment (PurpleBabar) [inline extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js]: Ok :) Will do :D

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js (hunk 6 above)
Comment (PurpleBabar) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js]: Oopsie, didn't pay attention, sorry :/

**Comment by aghassemi** [comment]
Thanks again @PurpleBabar, merged. This change will make it to Dev Chanel today and expected to be in production mid next week.

**Comment by PurpleBabar** [comment]
Thanks to you guys :D Thanks for the help and the patience :)

**Pull request merged**